cwms-data-api/src/main/java/cwms/cda/data/dto/location/kind/ProjectStructure.java

```diff
@@ -1,53 +1,76 @@
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
  */
 
 package cwms.cda.data.dto.location.kind;
 
 import com.fasterxml.jackson.annotation.JsonInclude;
 import com.fasterxml.jackson.databind.PropertyNamingStrategies;
 import com.fasterxml.jackson.databind.annotation.JsonNaming;
-import cwms.cda.api.errors.FieldException;
+import cwms.cda.api.errors.RequiredFieldException;
 import cwms.cda.data.dto.CwmsDTOBase;
-import cwms.cda.data.dto.Location;
 import cwms.cda.data.dto.CwmsId;
+import cwms.cda.data.dto.Location;
+import java.util.ArrayList;
+import java.util.List;
 
 @JsonInclude(JsonInclude.Include.NON_NULL)
 @JsonNaming(PropertyNamingStrategies.KebabCaseStrategy.class)
-public abstract class ProjectStructure implements CwmsDTOBase
-{
-	private final CwmsId projectId;
-	private final Location location;
-
-	protected ProjectStructure(CwmsId projectId, Location location)
-	{
-		this.location = location;
-		this.projectId = projectId;
-	}
-
-	public final CwmsId getProjectId()
-	{
-		return projectId;
-	}
-
-	public final Location getLocation()
-	{
-		return location;
-	}
-
-	@Override
-	public void validate() throws FieldException
-	{
-		if (getLocation() == null) {
-			throw new FieldException("Location field can't be null");
-		}
-		getLocation().validate();
-		if (getProjectId() == null) {
-			throw new FieldException("Project location Id field must be defined");
-		}
-		getProjectId().validate();
-	}
+public abstract class ProjectStructure implements CwmsDTOBase {
+    private final CwmsId projectId;
+    private final Location location;
+
+    ProjectStructure(CwmsId projectId, Location location) {
+        this.location = location;
+        this.projectId = projectId;
+    }
+
+    public final CwmsId getProjectId() {
+        return projectId;
+    }
+
+    public final Location getLocation() {
+        return location;
+    }
+
+    protected final List<String> getMissingFields() {
```

Comment:
if you make this non-final subclasses can extend and add to the list. Then the validate method could move here and be marked final.

Reply:
Would we want to force subclasses to implement getMissingFields()? (i.e. abstract the method and do the loc/parentId checks in validate?)  Or just de-finalize it, and let it be overridden as needed?

Reply:
I think in a refactor (out of scope for this PR) we'd make `getMissingFields()` abstract so that all DTO's override. Then the validate() can be consistent across.

Reply:
for the two concrete implementations of ProjectStructure so far I think it's ok not to abstract the getMissingFields()